openfl/utilities/optimizers/keras/fedprox.py

```diff
@@ -24,123 +23,51 @@ class FedProxOptimizer(keras.optimizers.Optimizer):
-        """
-        lr_t = tf.cast(self._lr_t, var.dtype.base_dtype)
-        mu_t = tf.cast(self._mu_t, var.dtype.base_dtype)
-        vstar = self.get_slot(var, "vstar")
-
-        var_update = var.assign_sub(lr_t * (grad + mu_t * (var - vstar)))
-
-        return tf.group(
-            *[
-                var_update,
-            ]
-        )
-
-    def _apply_sparse_shared(self, grad, var, indices, scatter_add):
-        """Apply sparse gradients to variables.
-
-        Args:
-            grad: Gradients.
-            var: Variables.
-            indices: A tensor of indices into the first dimension of `var`.
-            scatter_add: A scatter operation.
-
-        Returns:
-            A tf.Operation that applies the specified gradients.
-        """
-        lr_t = tf.cast(self._lr_t, var.dtype.base_dtype)
-        mu_t = tf.cast(self._mu_t, var.dtype.base_dtype)
-        vstar = self.get_slot(var, "vstar")
-        v_diff = vstar.assign(mu_t * (var - vstar), use_locking=self._use_locking)
-
-        with tf.control_dependencies([v_diff]):
-            scaled_grad = scatter_add(vstar, indices, grad)
-        var_update = var.assign_sub(lr_t * scaled_grad)
-
-        return tf.group(
-            *[
-                var_update,
-            ]
+    def __init__(
+        self,
+        learning_rate=0.01,
+        mu=0.0,
+        name="FedProxOptimizer",
+        **kwargs,
+    ):
+        super().__init__(
+            learning_rate=learning_rate,
+            name=name,
+            **kwargs,
         )
+        self.mu = mu
 
-    def _resource_apply_sparse(self, grad, var):
-        """
-        Apply sparse gradients to variables.
+    def build(self, variables):
+        """Initialize optimizer variables.
 
         Args:
-            grad: Gradients.
-            var: Variables.
-
-        Returns:
-            A tf.Operation that applies the specified gradients.
+          var_list: list of model variables to build FedProx variables on.
         """
-        return self._apply_sparse_shared(grad.values, var, grad.indices, standard_ops.scatter_add)
+        if self.built:
+            return
+        super().build(variables)
+        self.vstars = []
+        for variable in variables:
+            self.vstars.append(
+                self.add_variable_from_reference(
+                    reference_variable=variable, name="vstar"
+                )
+            )
+
+    def update_step(self, gradient, variable, learning_rate):
+        """Update step given gradient and the associated model variable."""
+        lr_t = tf.cast(learning_rate, variable.dtype)
+        mu_t = tf.cast(self.mu, variable.dtype)
+        gradient_t = tf.cast(gradient, variable.dtype)
+        vstar = self.vstars[self._get_variable_index(variable)]
+
+        self.assign_sub(variable, lr_t * (gradient_t + mu_t * (variable - vstar)))
 
     def get_config(self):
-        """Return the config of the optimizer.
```

Comment:
We should keep the docstring (and modify it if needed) for documentation purposes